From 1ce7afad47693212e3b216f02a1fa9c3ee621346 Mon Sep 17 00:00:00 2001
From: Khang D Nguyen <khangng@os.amperecomputing.com>
Date: Tue, 6 May 2025 09:42:12 +0700
Subject: [PATCH 4/4] Always use pipe for username and password

There was a bug with the method signature, see [1]. Frontend code was
changed but not the backend code, because it is not upstreamed yet.

Mount method now always takes in an Unix FD.

[1] https://github.com/openbmc/bmcweb/commit/4bbf2a13c21de665aab76ae3d3a3e2633b667be2

Upstream-Status: Inactive-Upstream
Signed-off-by: Khang D Nguyen <khangng@os.amperecomputing.com>
---
 src/state/initial_state.hpp | 74 ++++++++++++++++++-------------------
 1 file changed, 35 insertions(+), 39 deletions(-)

diff --git a/src/state/initial_state.hpp b/src/state/initial_state.hpp
index f9255b0..c806f03 100644
--- a/src/state/initial_state.hpp
+++ b/src/state/initial_state.hpp
@@ -18,7 +18,7 @@ struct InitialState : public BasicStateT<InitialState>
     }
 
     InitialState(interfaces::MountPointStateMachine& machine) :
-        BasicStateT(machine){};
+        BasicStateT(machine) {};
 
     std::unique_ptr<BasicState> handleEvent(RegisterDbusEvent event)
     {
@@ -223,55 +223,51 @@ struct InitialState : public BasicStateT<InitialState>
         if (isLegacy)
         {
             using sdbusplus::message::unix_fd;
-            using optional_fd = std::variant<int, unix_fd>;
 
             iface->register_method(
-                "Mount", [&machine = machine](boost::asio::yield_context yield,
-                                              std::string imgUrl, bool rw,
-                                              optional_fd fd) {
+                "Mount",
+                [&machine = machine](boost::asio::yield_context yield,
+                                     std::string imgUrl, bool rw, unix_fd fd) {
                     LogMsg(Logger::Info, "[App]: Mount called on ",
                            getObjectPath(machine), machine.getName());
 
                     interfaces::MountPointStateMachine::Target target = {
                         imgUrl, rw, nullptr, nullptr};
 
-                    if (std::holds_alternative<unix_fd>(fd))
+                    LogMsg(Logger::Debug, "[App] Extra data available");
+
+                    // Open pipe and prepare output buffer
+                    boost::asio::posix::stream_descriptor secretPipe(
+                        machine.getIoc(), dup(fd.fd));
+                    std::array<char, utils::secretLimit> buf;
+
+                    // Read data
+                    auto size = secretPipe.async_read_some(
+                        boost::asio::buffer(buf), yield);
+
+                    // Validate number of NULL delimiters, ensures
+                    // further operations are safe
+                    auto nullCount =
+                        std::count(buf.begin(), buf.begin() + size, '\0');
+                    if (nullCount != 2)
                     {
-                        LogMsg(Logger::Debug, "[App] Extra data available");
-
-                        // Open pipe and prepare output buffer
-                        boost::asio::posix::stream_descriptor secretPipe(
-                            machine.getIoc(), dup(std::get<unix_fd>(fd).fd));
-                        std::array<char, utils::secretLimit> buf;
-
-                        // Read data
-                        auto size = secretPipe.async_read_some(
-                            boost::asio::buffer(buf), yield);
-
-                        // Validate number of NULL delimiters, ensures
-                        // further operations are safe
-                        auto nullCount =
-                            std::count(buf.begin(), buf.begin() + size, '\0');
-                        if (nullCount != 2)
-                        {
-                            throw sdbusplus::exception::SdBusError(
-                                EINVAL, "Malformed extra data");
-                        }
-
-                        // First 'part' of payload
-                        std::string user(buf.begin());
-                        // Second 'part', after NULL delimiter
-                        std::string pass(buf.begin() + user.length() + 1);
-
-                        // Encapsulate credentials into safe buffer
-                        target.credentials =
-                            std::make_unique<utils::CredentialsProvider>(
-                                std::move(user), std::move(pass));
-
-                        // Cover the tracks
-                        utils::secureCleanup(buf);
+                        throw sdbusplus::exception::SdBusError(
+                            EINVAL, "Malformed extra data");
                     }
 
+                    // First 'part' of payload
+                    std::string user(buf.begin());
+                    // Second 'part', after NULL delimiter
+                    std::string pass(buf.begin() + user.length() + 1);
+
+                    // Encapsulate credentials into safe buffer
+                    target.credentials =
+                        std::make_unique<utils::CredentialsProvider>(
+                            std::move(user), std::move(pass));
+
+                    // Cover the tracks
+                    utils::secureCleanup(buf);
+
                     machine.emitMountEvent(std::move(target));
 
                     return true;
